connectors/cli.py

```diff
@@ -107,7 +107,12 @@ async def _run_service(service):
         else:
             return await service.run()
 
-    await asyncio.gather(*[_run_service(service) for service in services])
+    done, pending = await asyncio.wait(
+        [_run_service(service) for service in services],
+        return_when=asyncio.FIRST_COMPLETED,
```

Comment:
```suggestion
        return_when=asyncio.FIRST_EXCEPTION,
```

Should we use `FIRST_EXCEPTION` instead? 

When shutdown happens, it will call `_shutdown` to stop all services. when the first service stops completely, `asyncio.wait` will return, and here it will try to cancel the rest services, which are still completing.

Reply:
cc @tarekziade what do you think about it?

I've added tests to test both cases - when service exits with an exception or exits normally, and FIRST_COMPLETED works well, but I'm not 100% sure about it here.

Reply:
the documentation says:

- `FIRST_COMPLETED` - The function will return when any future finishes or is cancelled
- `FIRST_EXCEPTION` - The function will return when any future finishes by raising an exception. If no future raises an exception then it is equivalent to ALL_COMPLETED.

So yeah `COMPLETED` is valid.

here a full working example:

```
import asyncio

async def one():
    await asyncio.sleep(2)


async def two():
    await asyncio.sleep(1)
    raise Exception("two failed")

async def run():
    t1 = asyncio.create_task(one())
    t2 = asyncio.create_task(two())
    done, pending = await asyncio.wait([t1, t2], return_when=asyncio.FIRST_COMPLETED)

    for t in pending:
        t.cancel()
        try:
            await t
        except asyncio.CancelledError:
            pass

    for t in done:
        exc = t.exception()
        # raise first one
        if exc is not None:
            raise exc

    return done, pending

print(asyncio.run(run()))
```

Reply:
All the services in this repo are long-running services, so there are only 2 ways of existing:
1. service shutdown (expected)
2. some error happens (unexpected)

If it's `FIRST_EXCEPTION`:
For scenario 1), it's equivalent to `ALL_COMPLETED`, and `pending` will be empty, so it's no-op.
For scenario 2), when any task raise error, `asyncio.wait` will return and all the `pending` taks will be canceled.

If it's `FIRST_COMPLETED`:

For scenario 1), all tasks will be canceled (by setting `running` to `false`), and tasks will complete at different times. The first completed task will cause the rest to be `canceled`, I don't think this is desired.
For scenario 2), I assume raising an exception is not considered `completed`, so it will leave the rest running?

That's why I feel `FIRST_EXCEPTION` is what we needed here. But all these analyses are based on the `asyncio.wait` doc, and I have never tested it.

Reply:
`FIRST_COMPLETED` is triggered on exception, where an example with your two scenarios:


```
import asyncio

async def one():
    await asyncio.sleep(2)


async def two():
    await asyncio.sleep(1)
    raise Exception("two failed")

async def run(services):
    tasks = [asyncio.create_task(s) for s in services]
    done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)

    for t in pending:
        t.cancel()
        try:
            await t
        except asyncio.CancelledError:
            pass

    for t in done:
        exc = t.exception()
        # raise first one
        if exc is not None:
            raise exc

    return done, pending


# scenario 1
scenario_1 = one(), two()
try:
    print(asyncio.run(run(scenario_1)))
except Exception as e:
    print(e)

# scenario 2
scenario_2 = one(), one()
try:
    print(asyncio.run(run(scenario_2)))
except Exception:
    print(e)
```

Reply:
If you run this, you will see under a normal shutdown, only the first service completes successfully and the rest are canceled:
```python
import asyncio

class BaseService:
    def __init__(self):
        self.running = False

    async def run(self):
        print(f"Start {self.__class__.__name__}")
        self.running = True
        while self.running:
            await self._run()
        print(f"Complete {self.__class__.__name__}")

    def stop(self):
        print(f"Stop {self.__class__.__name__}")
        self.running = False

class ServiceOne(BaseService):
    async def _run(self):
        await asyncio.sleep(2)

class ServiceTwo(BaseService):
    async def _run(self):
        await asyncio.sleep(6)


async def run(services):
    tasks = [asyncio.create_task(s.run()) for s in services]
    # simulate SIGTERM to stop all services
    def _stop_services():
        for s in services:
            s.stop()
    asyncio.get_event_loop().call_later(1, _stop_services)
    done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)

    for t in pending:
        t.cancel()
        try:
            await t
        except asyncio.CancelledError:
            pass

    for t in done:
        exc = t.exception()
        # raise first one
        if exc is not None:
            raise exc

    return done, pending


services = [ServiceOne(), ServiceTwo()]
try:
    print(asyncio.run(run(services)))
except Exception as e:
    print(e)

```

You will see this:
```shell
chenhui.wang@SG-Chenhui Downloads % python3 asyncio_test.py
Start ServiceOne
Start ServiceTwo
Stop ServiceOne
Stop ServiceTwo
Complete ServiceOne
({<Task finished name='Task-2' coro=<BaseService.run() done, defined at /Users/chenhui.wang/Downloads/asyncio_test.py:7> result=None>}, {<Task cancelled name='Task-3' coro=<BaseService.run() done, defined at /Users/chenhui.wang/Downloads/asyncio_test.py:7>>})
```

Reply:
- you did not fully simulate SIGTERM in your example
- `ServiceTwo` gets properly stopped, but we just ignore the `CancelledError`in the example

If you want to have a `Complete` called on cancellation, you need to add it.

Example:

```
import asyncio
import os
import signal


class BaseService:
    def __init__(self):
        self.running = False

    async def run(self):
        print(f"Start {self.__class__.__name__}")
        try:
            self.running = True
            while self.running:
                await self._run()
        finally:
            print(f"Complete {self.__class__.__name__}")

    def stop(self):
        print(f"Stop {self.__class__.__name__}")
        self.running = False

class ServiceOne(BaseService):
    async def _run(self):
        await asyncio.sleep(2)

class ServiceTwo(BaseService):
    async def _run(self):
        await asyncio.sleep(6)


async def run(services):

    def _shutdown():
        for s in services:
            s.stop()

    asyncio.get_event_loop().add_signal_handler(signal.SIGTERM, _shutdown)

    tasks = [asyncio.create_task(s.run()) for s in services]

    # simulate SIGTERM to stop all services
    def _stop_services():
        os.kill(os.getpid(), signal.SIGTERM)

    asyncio.get_event_loop().call_later(1, _stop_services)
    done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)

    for t in pending:
        t.cancel()
        try:
            await t
        except asyncio.CancelledError:
            pass

    for t in done:
        exc = t.exception()
        # raise first one
        if exc is not None:
            raise exc

    return done, pending


services = [ServiceOne(), ServiceTwo()]
try:
    print(asyncio.run(run(services)))
except Exception as e:
    print(e)

```

One pattern you can have is to add a `tear_down` method `BaseService` for cleanups if you want

Reply:
example of integrating cancellation at the base class level:

```
class BaseService:
    def __init__(self):
        self.running = False

    async def run(self):
        print(f"Start {self.__class__.__name__}")
        try:
            self.running = True
            while self.running:
                await self._run()

            print(f"Complete {self.__class__.__name__}")
        except asyncio.CancelledError:
            self.cancelled()

    def _canceled(self):
        pass.  # can be overridden by concrete implementations

    def cancelled(self):
        print(f"Cancelled {self.__class__.__name__}")
        self._cancelled()

    def stop(self):
        print(f"Stop {self.__class__.__name__}")
        self.running = False
```